### PR TITLE
Bootstrap

### PR DESCRIPTION
**Pull request by vasilievip**

### CODE REVIEW

**Comment by vasilievip** [comment]
squash commits into one

File: .gitignore [hunk 1]
```diff
@@ -33,15 +33,12 @@ gen/
 # Local configuration file (sdk path, etc)
```
Comment (vasilievip) [inline .gitignore]: configure global gitignore

File: build.gradle [hunk 2]
```diff
@@ -48,6 +46,9 @@ dependencies {
     compile group: 'org.apache.commons', name: 'commons-dbcp2', version: '2.1.1'
     compile group: 'org.slf4j', name: 'slf4j-api', version: '1.7.21'
     compile group: 'org.apache.logging.log4j', name: 'log4j-core', version: '2.6.2'
+    compile group: 'org.webjars', name: 'bootstrap', version: '3.3.7-1'
+//    compile group: 'com.fasterxml.jackson.core', name: 'jackson-core', version: '2.8.2'
```
Comment (vasilievip) [inline build.gradle]: remove commented lines

File: src/main/java/application/Start.java [hunk 3]
```diff
@@ -1,8 +1,8 @@
 package application;
 
 
-import org.springframework.boot.SpringApplication;
-import org.springframework.boot.autoconfigure.SpringBootApplication;
+        import org.springframework.boot.SpringApplication;
```
Comment (vasilievip) [inline src/main/java/application/Start.java]: what is changed here?

File: build.gradle [hunk 4]
```diff
@@ -16,8 +16,6 @@ sonarqube {
     }
 }
 
-
-
 sourceCompatibility = 1.7
```
Comment (vasilievip) [inline build.gradle]: use 1.8 java

File: src/main/java/application/services/DefaultMenuService.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+
+    @Autowired
+    public DefaultMenuService(DishesRepository repository) {
+        this.repository = repository;
+    }
+
+    @Override
+    public List<Dish> getMenu() {
+        List<Dish> dishList = repository.findAll();
```
Comment (vasilievip) [inline src/main/java/application/services/DefaultMenuService.java]: return repository.findAll()

File: src/main/java/application/services/DefaultMenuService.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        return dishList;
+    }
+
+    @Override
+    public List<Dish> getSortedByPriceMenu() {
+        return new ArrayList<>();
```
Comment (vasilievip) [inline src/main/java/application/services/DefaultMenuService.java]: not implemented?

File: src/main/java/application/services/DefaultMenuService.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+
+    @Override
+    public Dish findByName(String name) {
+        return repository.findByDishNameIgnoreCase(name);
+
+    }
+
+    public List<Dish> performSortedSearch(String sorter) {
```
Comment (vasilievip) [inline src/main/java/application/services/DefaultMenuService.java]: if condition need to be in controller, service should provide 2 api calls

File: src/main/java/application/services/DefaultMenuService.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        if ("price".equals(sorter)) {
+            return repository.findAllByOrderByPriceAsc();
+        }
+        if ("description".equals(sorter)) {
+            return repository.findAllByOrderByDishNameAsc();
+        }
+        return new ArrayList<>();
+    }
+
+    public Order getOrderInformation(HttpServletRequest requestWithOrderInformation, Client client) {
```
Comment (vasilievip) [inline src/main/java/application/services/DefaultMenuService.java:48]: HttpServletRequest should stay in controller, service should have clean api with types parameters

File: src/main/java/application/services/DefaultOrderProcessingService.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@Component("orderService")
+public class DefaultOrderProcessingService implements OrderProcessingService {
+
+
+    List<Order> orders = new ArrayList<>();
```
Comment (vasilievip) [inline src/main/java/application/services/DefaultOrderProcessingService.java:20]: service should be stateless, remove state

Comment (vasilievip) [inline src/main/java/application/services/DefaultOrderProcessingService.java:20]: you can keep state in httpsession

File: src/main/java/application/services/DefaultOrderProcessingService.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+            }
+
+        }
+
+        return totalOrdersDTO;
+    }
+
+    @Override
+    public List<Order> getAllOrders() {
+        return orders;
```
Comment (vasilievip) [inline src/main/java/application/services/DefaultOrderProcessingService.java:53]: keep state in http session (or create session scoped beans to keep state)

File: src/main/java/application/services/DefaultOrderProcessingService.java [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    @Override
+    public void saveOrder(Order order) {
```
Comment (vasilievip) [inline src/main/java/application/services/DefaultOrderProcessingService.java:58]: http://stackoverflow.com/a/29416890

File: src/main/java/application/services/DefaultOrderProcessingService.java [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+            for(Map.Entry<Dish, Integer> d: o.getDishes().entrySet())
+            {
+              totalClientSum+=d.getValue()*d.getKey().getPrice();
+            }
+            clientOrderMap.put(o.getClient().getName(), totalClientSum);
+        }
+        return clientOrderMap;
+    }
+
+    private void addInitialOrder()
```
Comment (vasilievip) [inline src/main/java/application/services/DefaultOrderProcessingService.java:88]: wtf?

File: src/main/java/application/web/CartController.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+public class CartController {
+
+    @Autowired
+    DefaultMenuService service;
+
+    @RequestMapping(value = "/viewCart")
+    public ModelAndView showCart(HttpServletRequest request, @ModelAttribute Client client) {
+
+
+        ModelAndView mav = new ModelAndView("cart");
```
Comment (vasilievip) [inline src/main/java/application/web/CartController.java]: http://stackoverflow.com/a/29416890

File: src/main/java/application/web/model/TotalOrdersDTO.java [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class TotalOrdersDTO {
+
+    @Autowired
+    MenuService service;
```
Comment (vasilievip) [inline src/main/java/application/web/model/TotalOrdersDTO.java]: https://en.wikipedia.org/wiki/Data_transfer_object
Please read ftp://ftp.heanet.ie/mirrors/sourceforge/w/we/webtune/Patterns%20of%20Enterprise%20Application%20Architecture.pdf

File: src/main/java/application/web/model/PageDTO.java [hunk 15]
```diff
@@ -0,0 +1,8 @@
+package application.web.model;
+
```
Comment (vasilievip) [inline src/main/java/application/web/model/PageDTO.java]: wtf?

File: src/main/java/application/web/WelcomeController.java [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+    @RequestMapping("/")
+    public ModelAndView printWelcome() {
+        ModelAndView mav = new ModelAndView("index");
+        mav.addObject("client", new Client());
+        return mav;
+    }
+
+
+    @Override
+    protected void finalize() throws Throwable {
```
Comment (vasilievip) [inline src/main/java/application/web/WelcomeController.java:44]: https://github.com/HackathonHackers/programming-ebooks/blob/master/Java/Effective%20Java%20(2nd%20Edition).pdf
chapter 2

File: src/main/java/application/web/MenuController.java [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+
+    @Autowired
+    DefaultMenuService service;
+
+    @RequestMapping("/showMenu")
+    public ModelAndView requestHandlingMethod(HttpServletRequest request, @ModelAttribute Client client) {
+        Optional<String> sorter = null;
+        sorter = Optional.ofNullable(request.getParameter("sorter"));
+
+        List<Dish> menu;
```
Comment (vasilievip) [inline src/main/java/application/web/MenuController.java:32]: switch to java 8

List<Dish> menu = sorter.map( s -> service.performSortedSearch(sorter.get())).orElse(menu = service.getMenu());

File: src/main/java/application/web/MenuController.java [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+@SessionAttributes(types = Client.class)
+public class MenuController {
+
+    @Autowired
+    DefaultMenuService service;
+
+    @RequestMapping("/showMenu")
+    public ModelAndView requestHandlingMethod(HttpServletRequest request, @ModelAttribute Client client) {
+        Optional<String> sorter = null;
+        sorter = Optional.ofNullable(request.getParameter("sorter"));
```
Comment (vasilievip) [inline src/main/java/application/web/MenuController.java:30]: Optional<String> sorter = Optional.ofNullable(request.getParameter("sorter"));

File: src/main/java/application/web/OdrerProcessingController.java [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+
+    @Autowired
+    DefaultMenuService service;
+
+    @Autowired
+    OrderProcessingService orderService;
+
+    @RequestMapping(value = "/processOrder")
+    public String prosesOrder(@ModelAttribute Client client, HttpServletRequest request) {
+        Order order = (Order) request.getSession().getAttribute("order");
```
Comment (vasilievip) [inline src/main/java/application/web/OdrerProcessingController.java:37]: what is order is null?

File: src/main/java/application/web/CartController.java [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+        try {
+            Order order = service.getOrderInformation(request, (Client)request.getSession().getAttribute("client"));
```
Comment (vasilievip) [inline src/main/java/application/web/CartController.java]: input processing (wrong input) needs to stay in controller
1. http://howtodoinjava.com/best-practices/java-exception-handling-best-practices/
2. https://www.mkyong.com/spring-mvc/spring-mvc-exceptionhandler-example/

File: src/main/java/application/services/DefaultMenuService.java (hunk 7 above)
Comment (VitaliyMi) [reply src/main/java/application/services/DefaultMenuService.java]: Ok, but if we add a new feature, for example dishRate and we want to perform sorting by rate, we should expand our API? In our case, we can just add 1 procedure to performSortedSerch method. Isn't it useful for scalability of the project?

File: src/main/java/application/services/DefaultOrderProcessingService.java (hunk 12 above)
Comment (VitaliyMi) [reply src/main/java/application/services/DefaultOrderProcessingService.java:88]: This is my caprice. Let it stay there for a while. This helps the first person after project deployment understand what heppens in the result table.

File: src/main/java/application/web/WelcomeController.java (hunk 16 above)
Comment (VitaliyMi) [reply src/main/java/application/web/WelcomeController.java:44]: " This means that you should never do anything time-critical in a finalizer."
I'm trying to log in heroku the time, when gc destroies Controller object.

File: src/main/java/application/web/OdrerProcessingController.java (hunk 19 above)
Comment (VitaliyMi) [reply src/main/java/application/web/OdrerProcessingController.java:37]: Nicht fershtein

**Pull request merged**